### ISSUE TITLE
Empty bytes from storage and memory are encoded differently

### ISSUE THREAD

**Issue by dk1a** [foundry-rs/foundry#5811]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (5cfed63 2023-09-11T00:24:34.136213513Z)

### What command(s) is the bug in?

forge script, forge test

### Operating System

Linux

### Describe the bug

```solidity
// SPDX-License-Identifier: MIT
pragma solidity >=0.8.0;

import { Script } from "forge-std/Script.sol";

contract TestEncodeScript is Script {
  bytes private bytesStorage;

  function run() public view returns (bytes memory a, bytes memory b) {
    bytes memory bytesMemory;

    a = abi.encode(bytesStorage);
    b = abi.encode(bytesMemory);
  }
}
```

running this script via `forge script TestEncode.s.sol` (or an equivalent thing via `forge test`) returns:
```
[⠒] Compiling...
No files changed, compilation skipped
Script ran successfully.
Gas used: 24728

== Return ==
a: bytes 0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
b: bytes 0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000
```

They should be equal. Running an equivalent contract via remix returns:
```
0:
bytes: a 0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000
1:
bytes: b 0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000
```

I assume forge somehow appends an extra `bytes32(0)` if empty storage is being encoded

**Comment by mds1** [comment]
Hmm @Evalir on macOS with `forge 0.2.0 (675a824 2023-09-08T00:23:49.878725000Z)` I am not able to reproduce this—both test and script environments give identical outputs for `a` and `b`

**Comment by Evalir** [comment]
Yep same @mds1 — can't seem to reproduce this on macOS?

**Comment by mds1** [comment]
I guess it could be a linux specific issue but that would be surprising. Maybe someone else can repro on linux

**Comment by dk1a** [comment]
my bad - I had solc 0.8.13 and hadn't noticed it (and it was around during my reproduction too, I should've done a clean forge init)
this is apparently a solidity bug that was fixed in 0.8.15
